File: scripts/key_references.py
"""Shared key-reference parsing and conservative library-link resolution."""
from __future__ import annotations

import re
import unicodedata
from collections import defaultdict
from dataclasses import dataclass

KEY_REFERENCE_ROLES = {
    "foundation", "method", "baseline", "dataset", "survey", "related_work"}


@dataclass(frozen=True)
class ReferenceIndexes:
    doi: dict[str, str]
    title: dict[str, str]


def normalized_doi(value: object) -> str:
    text = str(value or "").strip().lower()
    text = re.sub(r"^https?://(dx\.)?doi\.org/", "", text)
    return re.sub(r"^doi:\s*", "", text)


def normalized_title(value: object) -> str:
    text = unicodedata.normalize("NFKD", str(value or "").lower())
    return re.sub(r"[^a-z0-9]+", " ", text).strip()


def parse_key_references(value: object) -> list[dict]:
    if not isinstance(value, list):
        return []
    output = []
    for item in value[:8]:
        if not isinstance(item, dict) or not str(item.get("title", "")).strip():
            continue
        role = str(item.get("role", "")).strip()
        output.append({
            "title": str(item.get("title", "")).strip(),
            "doi": str(item.get("doi", "")).strip(),
            "year": item.get("year"),
            "role": role if role in KEY_REFERENCE_ROLES else "",
            "reason": re.sub(r"\s+", " ", str(item.get("reason", ""))).strip(),
            "status": "external",
            "linked_card": None,
        })
    return output


def build_reference_indexes(records: list[dict]) -> ReferenceIndexes:
    doi_candidates: dict[str, list[str]] = defaultdict(list)
    title_candidates: dict[str, list[str]] = defaultdict(list)
    for record in records:
        slug = str(record.get("slug", "")).strip()
        if not slug:
            continue
        doi = normalized_doi(record.get("doi"))
        title = normalized_title(record.get("title"))
        if doi:
            doi_candidates[doi].append(slug)
        if len(title) > 12:
            title_candidates[title].append(slug)
    return ReferenceIndexes(
        doi={
            key: slugs[0]
            for key, slugs in doi_candidates.items()
            if len(slugs) == 1
        },
        title={
            key: slugs[0]
            for key, slugs in title_candidates.items()
            if len(slugs) == 1
        },
    )


def resolve_key_references(
    value: object,
    indexes: ReferenceIndexes,
    current_slug: str = "",
) -> list[dict]:
    references = parse_key_references(value)
    for reference in references:
        doi = normalized_doi(reference["doi"])
        linked = indexes.doi.get(doi) if doi else None
        if linked is None and not doi:
            title = normalized_title(reference["title"])
            linked = indexes.title.get(title) if len(title) > 12 else None
        if linked and linked != current_slug:
            reference["status"] = "in_library"
            reference["linked_card"] = linked
    return references

File: webapp/lib/key-references.ts
import { normalizedDoi, normalizedTitle } from "./duplicates";
import {
  KEY_REFERENCE_ROLES,
  type KeyReference,
  type KeyReferenceRole,
} from "./types";

type LinkableCard = {
  slug: string;
  title: string;
  doi: string;
};

function optionalYear(value: unknown): number | null {
  if (value === null || value === undefined || value === "") return null;
  const year = Number(value);
  return Number.isInteger(year) && year >= 1000 && year <= 3000 ? year : null;
}

function oneLine(value: unknown, limit = 400): string {
  return String(value || "").replace(/\s+/g, " ").trim().slice(0, limit);
}

export function parseKeyReferences(value: unknown): KeyReference[] {
  if (!Array.isArray(value)) return [];
  return value.flatMap((item) => {
    if (!item || typeof item !== "object") return [];
    const record = item as Record<string, unknown>;
    const title = oneLine(record.title, 500);
    if (!title) return [];
    const role: KeyReferenceRole | "" = KEY_REFERENCE_ROLES.includes(record.role as KeyReferenceRole)
      ? (record.role as KeyReferenceRole)
      : "";
    return [{
      title,
      doi: oneLine(record.doi, 200),
      year: optionalYear(record.year),
      role,
      reason: oneLine(record.reason),
      status: "external" as const,
      linked_card: null,
    }];
  }).slice(0, 8);
}

export function linkKeyReferences(
  references: KeyReference[],
  cards: LinkableCard[],
  currentSlug?: string,
): KeyReference[] {
  const candidates = cards.filter((card) => card.slug !== currentSlug);
  return references.map((reference) => {
    const doi = normalizedDoi(reference.doi);
    let match: LinkableCard | undefined;
    if (doi) {
      match = candidates.find((card) => normalizedDoi(card.doi) === doi);
    }
    if (!match && !doi) {
      const title = normalizedTitle(reference.title);
      if (title.length > 12) {
        const titleMatches = candidates.filter(
          (card) => normalizedTitle(card.title) === title,
        );
        if (titleMatches.length === 1) match = titleMatches[0];
      }
    }
    return match
      ? { ...reference, status: "in_library", linked_card: match.slug }
      : { ...reference, status: "external", linked_card: null };
  });
}
